feat(layout): Moved the outfitter and shipyard "detail information" into a sidebar

ThrosturX opened the pull request:
**Feature:** This PR implements the feature request detailed and discussed in issue #500

## Feature Details
I moved the information from the main shop panels into a new sidebar next to the ship sidebar, and made the new sidebar independently scroll-able.

## UI Screenshots
![outfitter before](https://user-images.githubusercontent.com/1678645/90324237-9b600280-df5b-11ea-8d3a-ce8cba7b9d74.png)
![outfitter after](https://user-images.githubusercontent.com/1678645/90324236-98fda880-df5b-11ea-8268-ab1996afb886.png)
![shipyard before](https://user-images.githubusercontent.com/1678645/90324240-9dc25c80-df5b-11ea-917b-a335ad6945f6.png)
![shipyard after](https://user-images.githubusercontent.com/1678645/90324239-9c912f80-df5b-11ea-94cf-387d30e9b56a.png)


## Testing Done
Basic functional UI testing with many ships in the hangar.

## Performance Impact
N/A

_Note: This is my first PR so bear with me if I overlooked something. I realize that some methods need renaming and some code bits may need refactoring, but I'm hoping the community can guide me in that regard with constructive criticism. Meanwhile I hope the functionality can be assessed and tested so that I can make any necessary improvements._

Amazinite commented:
Perhaps the ship image used in the sidebar should be the top-down sprite instead of the shipyard thumbnail, as the shipyard thumbnail is already visible on the left side (#4312).

ThrosturX commented:
> Perhaps the ship image used in the sidebar should be the top-down sprite instead of the shipyard thumbnail, as the shipyard thumbnail is already visible on the left side (#4312).

Coverage for #4312 added as requested:

![ship](https://user-images.githubusercontent.com/1678645/90334610-69cc5300-dfbe-11ea-962b-f068917ea442.png)

Amazinite commented on source/ShopPanel.cpp:
```diff
@@ -345,14 +345,17 @@ void ShopPanel::DrawDetailsSidebar()
 		Point descPoint(outfPoint.X(), outfPoint.Y() + shipInfo.OutfitsHeight());
 
-		SpriteShader::Draw(shipSprite, center);
+		SpriteShader::Draw(shipSprite, center, spriteScale);
```
You should also swizzle the selected ship according to the player's swizzle.

ThrosturX replied:
Thank you, solved with https://github.com/endless-sky/endless-sky/pull/5233/commits/093065f2f31b83b9402a4b623fcec05722660824.

Zitchas commented:
While I kind of like the idea of having the ship & stats moved away from its current "expanded out from the bottom of the ship you clicked on" approach; I'm not sure squishing it all into a sidebar is really the best approach as it is currently implemented. As shown in the screenshots, that results in the text being heavily squished and in some cases (like that Kimek Spire) the description text vanishes altogether.

I think this highlights the need for some kind of a two-step system in the shipyard. The sidebar (as implemented here) is a good first step, but I think the data portrayed should be stripped down to a few basic stats (hull/shield, cargo space, fuel, stock speeds, # of guns, # turrets) with the ability to double click on the ship to pull up a larger panel that has *all* the info, along with both a thumbnail and top-down view. (or even better, an EVN style scene that showcases the ship)

beccabunny commented:
>  in some cases (like that Kimek Spire) the description text vanishes altogether
This sidebar is scroll able, so this is not an issue as far as I understand it

tehhowch commented on source/ShopPanel.cpp:
```diff
@@ -265,54 +266,136 @@ void ShopPanel::DrawSidebar()
+
+	// Draw this string representing the selection, centered in the details side panel,
+	// It will either be the outfit name or the ship model.
+	string selectedItem = "Nothing Selected";
```
Ideally there would be a single iteminfodisplay that you could pull these details from, but that's a larger refactor than needed here (one I have already started).

ThrosturX replied:
Since you have already started working on this, can we presume this change request resolved?

ThrosturX commented on source/ShopPanel.cpp:
```diff
@@ -899,11 +982,27 @@ double ShopPanel::Zone::ScrollY() const
-	
-	scroll = max(0., min(maximum, scroll - dy));
-	
+	double *scroll;
```
Resolved in https://github.com/endless-sky/endless-sky/pull/5233/commits/e46022164af6e1108db7a2ef0b575d16a5b979ec

ThrosturX commented on source/OutfitterPanel.cpp:
```diff
@@ -208,23 +208,6 @@ int OutfitterPanel::DetailWidth() const
 
 
 
-int OutfitterPanel::DrawDetails(const Point &center)
```
Resolved in https://github.com/endless-sky/endless-sky/pull/5233/commits/e46022164af6e1108db7a2ef0b575d16a5b979ec

tehhowch commented on source/PlayerInfoPanel.cpp:
```diff
@@ -370,7 +370,7 @@ bool PlayerInfoPanel::KeyDown(SDL_Keycode key, Uint16 mod, const Command &comman
 
 
-bool PlayerInfoPanel::Click(int x, int y, int clicks)
+bool PlayerInfoPanel::Click(int /* x */, int /* y */, int clicks)
```
Seems like you cherry-picked a commit from `master` ... To resolve conflicts/add commits from the base branch, please use `git merge` with a regular merge commit. I believe you will need to force-push to remove this commit from the PR.

ThrosturX replied:
Thank you, I removed it. I'll defer updating from master for now.

tehhowch commented on source/ShopPanel.h:
```diff
@@ -117,10 +120,13 @@ class ShopPanel : public Panel {
-	bool dragMain = true;
+	double maxSideShipScroll = 0.;
+	double maxSideDetailScroll = 0.;
+	bool hoverInfo = false;
```
Rather than use two booleans to indicate which panel is being hovered / manipulated, let's define a class enum with 3 values, e.g.
```c++
enum class ShopPane : int {
  Main,
  Sidebar,
  Info
};
```
and then we need only one variable of that type, e.g. `activePane`, whose value will be e.g. `ShopPane::Main`.

ThrosturX replied:
Thank you, refactored in 376eac29.

tux2603 commented:
Has this been tested on a low resolution display yet?

tehhowch commented:
Probably not. The "official" minimum resolution is 1024x768 fullscreen, but I think that is already too small for a handful of map panels.

I have mentioned a resolution I use previously in a different UI PR.. but off the top of my head I'd say we should ensure it looks ok in the neighborhood of 1100x800 at 100% UI zoom.

tux2603 commented:
Here are some screenshots at 1280x1024:
![image](https://user-images.githubusercontent.com/45047084/90596087-3525ea80-e1bc-11ea-87ae-217a7f5c8df9.png)
![image](https://user-images.githubusercontent.com/45047084/90596114-44a53380-e1bc-11ea-95d3-1bba1c434b86.png)

And 1024x768
![image](https://user-images.githubusercontent.com/45047084/90596262-90f07380-e1bc-11ea-9b3d-6f6fd0167a12.png)
![image](https://user-images.githubusercontent.com/45047084/90596283-9bab0880-e1bc-11ea-8dc2-c7866521a6b8.png)

The smaller ships, especially the fighters and drones, look a bit small. Would it be possible to add some padding to a ship if it is less than a certain height? Overall though, I'm really liking this change

tehhowch commented:
I also agree that there should be some minimum + maximum detail tile size, to minimize UI fluctuation as the player sifts through the available items.

There is also a now-unnecessary table underline drawn beneath the selected item in the main panel. This used to serve as the border of the details row.
(There's an issue that may get closed by this PR relating to that line: ...)

ThrosturX commented:
> There is also a now-unnecessary table underline drawn beneath the selected item in the main panel. This used to serve as the border of the details row.

How about this? Are we happy with this?

![noline](https://user-images.githubusercontent.com/1678645/90633970-8662c800-e216-11ea-95ae-235eab5887dc.png)

beccabunny commented:
I'd say remove the line entirely, and perhaps add it below the outfit image in the sidebar. An alternative to that would be adding the "outfit background" (the 3d grid used in the outfitter) also behind the sidebar sprite

ThrosturX commented:
> I'd say remove the line entirely, and perhaps add it below the outfit image in the sidebar. An alternative to that would be adding the "outfit background" (the 3d grid used in the outfitter) also behind the sidebar sprite

After some inspection I was unable to find any code that draws that line. Turns out it's a part of `ui/selected outfit` (and ship). I won't be changing any images in this PR but if the line is to be removed, that is how to go about it.

tehhowch commented:
It should be readily apparent to the user which outfit in the main panel is selected. In the current view, this is achieved with 4 elements: the name is brighter, the background grid is brighter, there is a large row gap beneath the item's row, and the text is centered under the item.

The name and grid changes are quite subtle, so we may wish to augment the selected tile further. One option is a lighter background color, indicating a "tile highlight," and another is a border around the selected item tile. The border is troublesome because outfits are intentionally allowed to "spill out," to visually suggest space requirements.

We don't have to get the selected item distinction perfect here, but it should be passable.

Amazinite commented:
![shipyard selected+@2x](https://user-images.githubusercontent.com/17688683/90715526-809dce80-e278-11ea-9e21-cdcc46216c32.png)
![outfitter selected+@2x](https://user-images.githubusercontent.com/17688683/90715527-809dce80-e278-11ea-976e-0036feff7bc9.png)
[assets.zip](https://github.com/endless-sky/endless-sky/files/5100598/assets.zip)

Both the 1x and 2x images are included in the assets zip folder for you to download and use for this PR. The 2x images need to be PRed to the [high DPI repository](https://github.com/endless-sky/endless-sky-high-dpi).

ThrosturX commented:
I updated the images and added some "tile shading", it looks like this:

![outf](https://user-images.githubusercontent.com/1678645/90809070-7deeb800-e310-11ea-824a-afed7d36de0a.png)
![ship](https://user-images.githubusercontent.com/1678645/90809081-7fb87b80-e310-11ea-8928-e1b8563f12a2.png)

Not perfect, ideally we could somehow re-work the selected ship/outfit images a bit (like remove the line at the top of selected ship and make the selected outfit grid lines much brighter). But I think it's passable. Are we happy with this?

Amazinite commented:
I’d say just go with the images I provided. No reason to also shade the background differently.

ThrosturX commented:
> I’d say just go with the images I provided. No reason to also shade the background differently, as I think the existing differences between unselected and selected items are enough.

Are you sure? I feel that there is hardly any difference between selected and selected without the fill... I've been playtesting on this PR for a couple of days and it was already hard with the now out-of-place line at the bottom still there. See screenshots.

![nofill](https://user-images.githubusercontent.com/1678645/90825386-5e638980-e328-11ea-8c5d-47a8c72cc674.png)
![noselect](https://user-images.githubusercontent.com/1678645/90825388-5e638980-e328-11ea-8c48-dfbbaa890470.png)
![fill](https://user-images.githubusercontent.com/1678645/90825382-5d325c80-e328-11ea-8204-f5bdd2a56eb3.png)

litrevan commented:
Strictly as a comment on history of resolution. EV and EVN were both designed to look better in the newly available  800x600, rather than the standard 640*480, and it really showed in playability and readability. Now we often have the opposite problem. Too large of a resolution, and can't read the small print, especially anybody older.

tehhowch commented on source/ShopPanel.h:
```diff
@@ -117,10 +124,12 @@ class ShopPanel : public Panel {
 	double maxMainScroll = 0.;
-	double maxSideScroll = 0.;
-	bool dragMain = true;
+	double maxSideShipScroll = 0.;
+	double maxSideDetailScroll = 0.;
```
same idea here

tehhowch commented on source/ShopPanel.h:
```diff
@@ -117,10 +124,12 @@ class ShopPanel : public Panel {
 	
 	double mainScroll = 0.;
-	double sideScroll = 0.;
+	double sideShipScroll = 0.;
+	double sideDetailScroll = 0.;
```
if renames are done, also do here.

tehhowch commented on source/ShopPanel.cpp:
```diff
@@ -411,34 +442,9 @@ void ShopPanel::DrawMain()
 			isEmpty = false;
 			if(isCollapsed)
 				break;
-			
+
```
restore

ThrosturX replied:
I think I fixed it, confirmed?

tehhowch commented on source/ShopPanel.h:
```diff
@@ -30,7 +30,11 @@ class Planet;
+  Main,
+  Sidebar,
+  Info
+};
 
 // Class representing the common elements of both the shipyard panel and the
```
Need another indented newline here

ThrosturX replied:
I think I fixed this too, confirmed?

ThrosturX commented on source/ShopPanel.cpp:
```diff
@@ -265,54 +266,84 @@ void ShopPanel::DrawSidebar()
+{
+	// Fill in the background.
+	const Color &line = *GameData::Colors().Get("dim");
+	const Color back = Color(.055, .055, .055);
```
Done, named `shop info panel background`.

ThrosturX commented on source/ShopPanel.h:
```diff
@@ -97,7 +102,9 @@ class ShopPanel : public Panel {
 protected:
-	static const int SIDE_WIDTH = 250;
+	static const int SHIP_SIDE_WIDTH = 250;
+	static const int INFO_SIDE_WIDTH = 300;
```
Renames done.

Amazinite commented on source/OutfitterPanel.cpp:
```diff
@@ -210,17 +209,49 @@ int OutfitterPanel::DetailWidth() const
+		const Sprite *thumbnail = selectedOutfit->Thumbnail();
+		if(!thumbnail)
+		{
+			thumbnail = SpriteSet::Get("ui/outfitter selected");
```
Should we always render the outfit selected image behind the outfit image instead of only when there is no image to show? And same deal with the ships; should we put the ship selected image behind the image of the ship? If you look at the right-most bar, you'll notice that the currently selected escort actually has the selected ship image behind it.

ThrosturX replied:
I honestly don't care how we do it. This way is fine, the other way is fine, but in all seriousness someone needs to make the call here. Without an actual PO (or PM) at the reins I think it boils down to the testers or at least the most senior developers to make the call, but in any case it is a super easy change to make. Just make the call, what do you think makes the most sense, because I have been play testing this and I definitely feel that it doesn't make a difference and a little input goes a long way here.

tehhowch reviewed (commented):
Let's always have the background, as Amazinite proposed.

Then from my [comment on discord](https://discord.com/channels/251118043411775489/285540320823869441/750692182314319905), let's put the outfit or ship description before any stats, just like it is currently, but make that description field collapsible so people who are done with lore don't have to scroll for numbers.

We can make other fields collapsible or experiment (e.g. an on-hover description) in future PRs.

ThrosturX commented:
I have made the changes as discussed here and in the discord. The description is once again first after the thumbnail and has been made collapsible. I decided to make use of the `categoryZones` code, favoring KISS, as I strongly suspect most of the code will be refactored into more generic UI code eventually.

Here is a quick recording of the functionality in action, encoded into the (unfortunately) only supported video format available:

![simplescreenrecorder-2020-09-06_18 53 03](https://user-images.githubusercontent.com/1678645/92333122-a4338800-f072-11ea-8485-763018404902.gif)

Amazinite commented:
Looks amazing. My only comment right now on the appearance would be that there's no indication that the description can be collapse. Looking at the ship or outfit categories as an example, they have the downward pointing arrow to show that the category is expanded, and clicking that causes it to point to the right to show that it is collapsed, but with the description right now there is no "expanded" arrow.

Ririshi commented:
Looks like a good addition to the game! Definitely makes the outfitter more tidy. Since you are working on reworking the outfit interface with a sidebar here, I was thinking for some time that it would be nice to have a separate sidebar for the currently selected ship. Seeing your ship's info can be particularly hard when you have lots of ships (like in the screenshots posted by @ThrosturX).  On the other hand, combining this PR's sidebar with yet another one may lead to very little room for the actual outfits, especially on smaller screens.

Alternatively, perhaps it would be interesting to move the list of ships into a bottom bar rather than a side bar, freeing up the space in the old sidebar to only show the currently selected ship's information.

![image](https://user-images.githubusercontent.com/8461226/92338329-0199fa00-f0b0-11ea-9e30-c32e383e70b9.png)

As you can see, there is a lot of space still unused on my 1440p monitor. Perhaps something like this will use up both the vertical and the horizontal space nicely? Took me a long time to make this in Paint.net but I'm sure someone can make this work nicely in the actual game.

![image](https://user-images.githubusercontent.com/8461226/92339239-5a1fc600-f0b5-11ea-9f9e-c2f69d5ce4fa.png)

Personally, I wouldn't mind having another sidebar on my screen size, but for smaller screens, it makes more sense to have a bottom bar with three or two rows of ships. Maybe it shouldn't extend throughout the width of the screen though. Any thoughts would be welcome :) I have no clue how to actually make this in the source, but I might try to fiddle with some stuff if I can find time between uni, work, and just enjoying the game :)

beccabunny commented:
> Looks like a good addition to the game! Definitely makes the outfitter more tidy. Since you are working on reworking the outfit interface with a sidebar here, I was thinking for some time that it would be nice to have a separate sidebar for the currently selected ship. Seeing your ship's info can be particularly hard when you have lots of ships (like in the screenshots posted by @ThrosturX). On the other hand, combining this PR's sidebar with yet another one may lead to very little room for the actual outfits, especially on smaller screens.
> 
> Alternatively, perhaps it would be interesting to move the list of ships into a bottom bar rather than a side bar, freeing up the space in the old sidebar to only show the currently selected ship's information.
> 
> ![image](https://user-images.githubusercontent.com/8461226/92338329-0199fa00-f0b0-11ea-9e30-c32e383e70b9.png)
> 
> As you can see, there is a lot of space still unused on my 1440p monitor. Perhaps something like this will use up both the vertical and the horizontal space nicely? Took me a long time to make this in Paint.net but I'm sure someone can make this work nicely in the actual game.
> 
> ![image](https://user-images.githubusercontent.com/8461226/92339239-5a1fc600-f0b5-11ea-9f9e-c2f69d5ce4fa.png)
> 
> Personally, I wouldn't mind having another sidebar on my screen size, but for smaller screens, it makes more sense to have a bottom bar with three or two rows of ships. Maybe it shouldn't extend throughout the width of the screen though. Any thoughts would be welcome :) I have no clue how to actually make this in the source, but I might try to fiddle with some stuff if I can find time between uni, work, and just enjoying the game :)
> 
> Edit: I thought of another design that is not as extreme a change as the other one, but still nicer to use than the current setup: put the ship details at the top, and move the "your ships" below it, Only screens with sufficient height would allow that, though, since you need at least 3 or 4 rows of ships to scroll through to make it anywhere near usable. Example (again on my 1440p screen):
> 
> ![image](https://user-images.githubusercontent.com/8461226/92339689-ba176c00-f0b7-11ea-862c-b07eddcd21f6.png)

Personally, I don't really see it as necessary night now; not only the space in small monitors is much smaller, even on the bottom, but seeing the ship's info is just as easy as scrolling, and this solution does not look much more intuitive to me

Ririshi commented:
> this solution does not look much more intuitive to me

What about the last picture though? I personally really dislike having to scroll all the way down to see my currently active ship. Keeping the active ship information pinned at the top would make the UI feel more robust to me.

The only issue to tackle then would be to support screens that have too little vertical space to show all of the ship information and still show a decent number of ships at the bottom. To remedy that, I suggest to make ship info and ship list into separate scrollable blocks. While building the outfitter interface, attempt to show the whole ship info, followed by the ship list. If the screen is too small to draw at least three (or two or four depending on design choice) rows of ships below it, make the shpi info scrollable. That way, you'll always be able to see your ship's statistics, while also having the ability to quickly select different ships.

Amazinite commented:
@Ririshi seems to me like you're looking to have #5143 resolved, while this PR is looking to solve #500 (and just so happens to also be able to solve #1807 and #4312), so I'd say we tackle that at another time.

Ririshi commented:
> so I'd say we tackle that at another time

Fair enough, sorry to take this discussion off-topic!

tehhowch commented on source/OutfitterPanel.cpp:
```diff
@@ -25,6 +24,7 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include "Planet.h"
 #include "PlayerInfo.h"
 #include "Point.h"
+#include "PointerShader.h"
```
Used?

ThrosturX replied:
Nix; nixed.

tehhowch commented on source/OutfitterPanel.cpp:
```diff
@@ -210,17 +210,83 @@ int OutfitterPanel::DetailWidth() const
+		const Sprite *thumbnail = selectedOutfit->Thumbnail();
+		const Sprite *background = SpriteSet::Get("ui/outfitter selected");
+
+		float tileSize = max(thumbnail->Height(), static_cast<float>(TileSize()));
```
Segfault here for outfits with no sprite

ThrosturX replied:
Fixed with ternary operator, but not tested.

tehhowch replied:
![image](https://user-images.githubusercontent.com/20871346/93345747-cb1f5600-f7f8-11ea-8714-9d970091aba4.png)
👍

tehhowch commented on source/OutfitterPanel.cpp:
```diff
@@ -210,17 +210,83 @@ int OutfitterPanel::DetailWidth() const
+
+		// Maintenance note: This can be replaced with collapsed.contains() in C++20
+		if(!collapsed.count("description"))
+		{
+			descriptionOffset = outfitInfo.DescriptionHeight();
+			outfitInfo.DrawDescription(startPoint);
+		}
```
Since it would seem strange and out-of-place to render the "expanded" pointer for this case, maybe we should highlight the description on hover -- that way the player can intuit that they can do something to the field

ThrosturX replied:
This has been added, using "bright" color.

tehhowch commented on source/OutfitterPanel.cpp:
```diff
@@ -210,17 +210,83 @@ int OutfitterPanel::DetailWidth() const
+		Point descDimensions(INFOBAR_WIDTH, descriptionOffset + 10.);
+		ClickZone<std::string> collapseDescription = ClickZone<std::string>(descCenter, descDimensions, std::string("description"));
+
+		// find the old zone to erase it
+		for(auto it = categoryZones.begin(); it != categoryZones.end(); ++it)
+		{
+			if(it->Value() == "description")
+			{
+				categoryZones.erase(it);
+				break;
+			}
+
+		}
+
+		// insert the new zone
+		categoryZones.emplace_back(collapseDescription);
```
nit: Can we overwrite the existing clickzone with the new one, rather than erase and then append?

ThrosturX replied:
I'm not very motivated to do this -- we'll always have to iterate anyway to check if the clickzone exists, because if it doesn't exist it needs to be added instead. I think it's easier to understand this way and since this isn't computationally expensive I don't see a reason to change the behavior.

tehhowch commented on source/ShopPanel.h:
```diff
@@ -30,8 +30,6 @@ class Planet;
 class PlayerInfo;
 class Ship;
 
-
-
```
Some styling lines got nixed when the enum was moved in throughout header; please restore them

ThrosturX replied:
Thank you, restored.

tehhowch reviewed (commented):
Looking great!

Some extra questions
1) Currently, the "expand all" and "collapse all" shift-clicks on category zones also affects the description zone -- should it? (I'm fine keeping this behavior as long as you are ok with it.)
2) Can we remove the changes to the line thicknesses in the shipyard selected sprite? (This is perhaps more for @Amazinite )

Amazinite commented:
> Can we remove the changes to the line thicknesses in the shipyard selected sprite?

I made the sprites here using the original .svg files. If there's any difference in appearance outside of what was intended, I can only imagine that it is a result of me using a newer Inkscape version than when the graphics were originally made (as Inkscape did ask me about the .svg files being from an older version).

(And I wouldn't say the shipyard selected sprite looks like the lines are thicker, just brighter, which given that one of the mentions concerns has been being able to tell where your selected ship/outfit is, I don't see that as a bad thing that the sprite is brighter.)

tehhowch reviewed (approved):
Still on the fence about the sprite (it looks like the lines are thicker to me), but otherwise LGTM!

tehhowch commented:
@Amazinite if there's no high-DPI update needed, then this is ready for merging

DingusShingleton commented:
I just noticed the new patch so I decided to hop back into ES after a few years. Really glad to see it's still humming along! Then I noticed the column of stats in the outfitter screen and it hit me.... I requested that! It was the first issue I ever opened, and now, 5 and a half years later, it's in the game. Wild! And thank you!

Now just to add a tiny piece of "useful" information to this comment, in observing the new layout, I never would have known that the description was collapsible (elegant fix btw), and I only discovered that it was when I read @tehhowch's comment in this pull request saying as much. So +1 for adding the little universal "this collapses" triangle. Otherwise, tremendous stuff.
